File: packages/go/locale.go
package locale

import (
	"embed"
	"encoding/json"
	"fmt"
	"strings"
)

//go:embed locales/*.json
var localeFS embed.FS

var cache = map[string]map[string]any{}

func GetLocaleData(locale string) map[string]any {
	effectiveLocale := locale
	if effectiveLocale == "" {
		effectiveLocale = "en"
	}
	if cached, ok := cache[effectiveLocale]; ok {
		return cached
	}
	payload, err := localeFS.ReadFile("locales/" + effectiveLocale + ".json")
	if err != nil {
		panic(err)
	}
	var data map[string]any
	if err := json.Unmarshal(payload, &data); err != nil {
		panic(err)
	}
	cache[effectiveLocale] = data
	return data
}

func Translate(key string, locale string, params map[string]any) (string, bool) {
	value, ok := resolve(GetLocaleData(locale), key)
	if !ok {
		value, ok = resolve(GetLocaleData("en"), key)
	}
	if !ok {
		return key, false
	}
	text, isString := value.(string)
	if !isString {
		return key, false
	}
	for name, replacement := range params {
		text = strings.ReplaceAll(text, "{"+name+"}", fmt.Sprint(replacement))
	}
	return text, true
}

func resolve(data map[string]any, key string) (any, bool) {
	var current any = data
	for _, part := range strings.Split(key, ".") {
		next, ok := current.(map[string]any)
		if !ok {
			return nil, false
		}
		current, ok = next[part]
		if !ok {
			return nil, false
		}
	}
	return current, true
}

File: packages/composer/src/Locale.php
<?php

declare(strict_types=1);

namespace Escalated\Locale;

final class Locale
{
    /** @var array<string, array<string, mixed>> */
    private static array $cache = [];

    /** @return array<string, mixed> */
    public static function getLocaleData(string $locale = 'en'): array
    {
        if (!isset(self::$cache[$locale])) {
            $path = dirname(__DIR__).'/locales/'.$locale.'.json';
            self::$cache[$locale] = json_decode((string) file_get_contents($path), true, 512, JSON_THROW_ON_ERROR);
        }

        return self::$cache[$locale];
    }

    /** @param array<string, scalar> $params */
    public static function translate(string $key, string $locale = 'en', array $params = []): string
    {
        $value = self::resolve(self::getLocaleData($locale), $key) ?? self::resolve(self::getLocaleData('en'), $key);
        if (!is_string($value)) {
            return $key;
        }

        foreach ($params as $name => $replacement) {
            $value = str_replace('{'.$name.'}', (string) $replacement, $value);
        }

        return $value;
    }

    /** @param array<string, mixed> $data */
    private static function resolve(array $data, string $key): mixed
    {
        $current = $data;
        foreach (explode('.', $key) as $part) {
            if (!is_array($current) || !array_key_exists($part, $current)) {
                return null;
            }
            $current = $current[$part];
        }

        return $current;
    }
}
